include/sampleflow/producers/delayed_rejection_mh.h

```diff
@@ -0,0 +1,164 @@
+
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2020 by the SampleFlow authors.
+//
+// This file is part of the SampleFlow library.
+//
+// The SampleFlow library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of SampleFlow.
+//
+// ---------------------------------------------------------------------
+
+#ifndef SAMPLEFLOW_PRODUCERS_DELAYED_REJECTION_MH_H
+#define SAMPLEFLOW_PRODUCERS_DELAYED_REJECTION_MH_H
+
+#include <sampleflow/producer.h>
+#include <sampleflow/scope_exit.h>
+#include <sampleflow/types.h>
+
+#include <random>
+#include <cmath>
+
+namespace SampleFlow
+{
+  namespace Producers
+  {
+    /**
+     * An implementation of Delayed Rejection Metropolis Hastings, or DRAM.
+     * This producer functions very similarly to the regular Metropolis Hastings
+     * producer - consult the documentation of that class for more details.
+     * This class provides the same functionality as Metropolis Hastings but with
+     * the additional capability to perform delayed rejection. This requries a slightly
+     * different perturb function that accepts both the current sample and a vector
+     * of rejected samples as arguments. These rejected samples can be used for a more
+     * sophisticated perturbation strategy.
+     */
+    template <typename OutputType>
+    class DelayedRejectionMetropolisHastings : public Producer<OutputType>
+    {
+      public:
+        /**
+         * The principal function of this class. Starting from the given
+         * initial sample $x_0$, it produces a sequence of samples $x_k$
+         * that are passed through the signal of the base class to
+         * Consumer objects.
+         *
+         * @param[in] starting_point The initial sample $x_0$.
+         * @param[in] log_likelihood A function object that, when called
+         *   with a sample $x$, returns $\log(\pi(x))$, i.e., the natural
+         *   logarithm of the likelihood function evaluated at the sample.
+         * @param[in] perturb A function object that, when given a sample
```

Comment:
I think you should update the description of this parameter for the DR case.

Reply:
@bangerth Would you please elaborate? Are you referring to `log_likelihood` or `perturb`? I'm not sure what needs to be changed about the description.